### PR TITLE
Add forceStop flag to provide facility to stop healing

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add forceStop flag to provide facility to stop healing

## Motivation and Context
This PR also makes sure that we deal with HTTP request
count by ignoring the on-going heal operation, i.e
do not wait on itself.

## Regression
No

## How Has This Been Tested?
Using `mc admin heal` changes here https://github.com/minio/mc/pull/2584

```
~ mc admin heal --recursive myminio-local1-ssl/
```

```
~ mc admin heal --force-stop myminio-local1-ssl/
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by vadmeste — DISMISSED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#ce3c5dad-fee1-487e-8440-ec57e429713f) |
| mint-gateway-nas.sh | :x: [more...](#be64fa3e-f6ed-4459-bc11-a0efec548265) |
| mint-large-bucket.sh | :x: [more...](#ee387086-4c69-45f7-9bad-7210a16ed1bb) |
| mint-fs.sh | :x: [more...](#657a1bdf-557a-45fc-8ca1-bac4fdce683e) |
| mint-dist-xl.sh | :x: [more...](#51434472-74b5-4d22-901a-a4ae83ba1d99) |
<a name="51434472-74b5-4d22-901a-a4ae83ba1d99"></a>
### 6718-c67380b/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30293
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9315c68ac6ca:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 11 seconds
{
  "name": "awscli",
  "duration": 4607,
  "function": "aws --endpoint-url http://minikube:30293 s3api head-object --bucket awscli-mint-test-bucket-16967 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="ce3c5dad-fee1-487e-8440-ec57e429713f"></a>
### 6718-c67380b/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30053
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp dfc862a87735:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 26 seconds
{
  "name": "awscli",
  "duration": 10731,
  "function": "aws --endpoint-url http://minikube:30053 s3api head-object --bucket awscli-mint-test-bucket-16942 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="be64fa3e-f6ed-4459-bc11-a0efec548265"></a>
### 6718-c67380b/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31807
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp daf93eaf8fb4:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 25 seconds
{
  "name": "awscli",
  "duration": 10635,
  "function": "aws --endpoint-url http://minikube:31807 s3api head-object --bucket awscli-mint-test-bucket-6145 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="657a1bdf-557a-45fc-8ca1-bac4fdce683e"></a>
### 6718-c67380b/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30136
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 671ec324c8f0:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 26 seconds
{
  "name": "awscli",
  "duration": 10337,
  "function": "aws --endpoint-url http://minikube:30136 s3api head-object --bucket awscli-mint-test-bucket-8233 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="ee387086-4c69-45f7-9bad-7210a16ed1bb"></a>
### 6718-c67380b/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31835
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 16d996b42cd4:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 14 seconds
{
  "name": "awscli",
  "duration": 5368,
  "function": "aws --endpoint-url http://minikube:31835 s3api head-object --bucket awscli-mint-test-bucket-7089 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#300f3e58-d139-48e7-8b43-8ce721a8d814) |
| mint-gateway-nas.sh | :x: [more...](#22115f55-a2d5-4225-bf96-b705dab9055d) |
| mint-large-bucket.sh | :x: [more...](#c4b64dd5-d264-48ad-96b4-7bf30cf32ccb) |
| mint-fs.sh | :x: [more...](#a7779b68-c84a-4e2b-bce0-88bf72d6a2ae) |
| mint-dist-xl.sh | :x: [more...](#c7386f37-86d4-47d0-b7d6-040c0d9bdefd) |
<a name="c4b64dd5-d264-48ad-96b4-7bf30cf32ccb"></a>
### 6718-445600b/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31951
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 426278810eeb:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 13 seconds
{
  "name": "awscli",
  "duration": 5559,
  "function": "aws --endpoint-url http://minikube:31951 s3api head-object --bucket awscli-mint-test-bucket-21227 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="a7779b68-c84a-4e2b-bce0-88bf72d6a2ae"></a>
### 6718-445600b/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31150
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 34eaa8999367:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 25 seconds
{
  "name": "awscli",
  "duration": 10513,
  "function": "aws --endpoint-url http://minikube:31150 s3api head-object --bucket awscli-mint-test-bucket-21080 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="300f3e58-d139-48e7-8b43-8ce721a8d814"></a>
### 6718-445600b/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30657
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d3be7c5a04ed:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 24 seconds
{
  "name": "awscli",
  "duration": 11159,
  "function": "aws --endpoint-url http://minikube:30657 s3api head-object --bucket awscli-mint-test-bucket-11402 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="c7386f37-86d4-47d0-b7d6-040c0d9bdefd"></a>
### 6718-445600b/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32119
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5b1ce1e77f6b:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 10 seconds
{
  "name": "awscli",
  "duration": 4939,
  "function": "aws --endpoint-url http://minikube:32119 s3api head-object --bucket awscli-mint-test-bucket-9310 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```
<a name="22115f55-a2d5-4225-bf96-b705dab9055d"></a>
### 6718-445600b/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31023
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0 
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5025bead0d5a:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 24 seconds
{
  "name": "awscli",
  "duration": 10295,
  "function": "aws --endpoint-url http://minikube:31023 s3api head-object --bucket awscli-mint-test-bucket-5681 --key prefix/directory\n",
  "status": "FAIL",
  "error": "An error occurred (404) when calling the HeadObject operation: Not Found"
}

Executed 0 out of 13 tests successfully.
```

**Review by vadmeste — DISMISSED** [review]
LGTM & tested

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/admin-handlers.go [hunk 1]
```diff
@@ -520,10 +523,15 @@ func (a adminAPIHandlers) HealHandler(w http.ResponseWriter, r *http.Request) {
...
+					if started {
+						w.Write(hr.respBytes)
+						w.(http.Flusher).Flush()
+					} else {
+						writeSuccessResponseJSON(w, hr.respBytes)
+					}
+				case hr.errBody == "" && !started:
 					writeErrorResponseJSON(w, hr.errCode, r.URL)
-				default:
+				case !started:
```
Comment (donatello) [inline cmd/admin-handlers.go]: Looks like the case when `started && hr.errCode != ErrNone` is not handled in this switch case? We would need to write error message without response headers to the http writer.

File: pkg/madmin/heal-commands.go [hunk 2]
```diff
@@ -196,8 +203,12 @@ func (adm *AdminClient) Heal(bucket, prefix string, healOpts HealOpts,
 		queryVals.Set("clientToken", clientToken)
 		body = []byte{}
 	}
+
+	// Anyone can be set, either force start or forceStop.
 	if forceStart {
 		queryVals.Set("forceStart", "true")
+	} else if forceStop {
+		queryVals.Set("forceStop", "true")
```
Comment (donatello) [inline pkg/madmin/heal-commands.go:215]: I think we can help programming errors to be caught sooner by refusing to send the request when both are set.

**Review by donatello — DISMISSED** [review]

File: cmd/admin-handlers.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: That would be incorrect - client code will become complex.  Also this is a one time operation the actual status per heal is written by a different code path.

File: pkg/madmin/heal-commands.go (hunk 2 above)
Comment (harshavardhana) [reply pkg/madmin/heal-commands.go:215]: @donatello that can be done.

File: cmd/admin-handlers.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: Actually you may be right, need to add an extra `if` to make it work. But need to change client code as well to handle it.

Comment (harshavardhana) [reply cmd/admin-handlers.go]: Done

File: pkg/madmin/heal-commands.go (hunk 2 above)
Comment (harshavardhana) [reply pkg/madmin/heal-commands.go:215]: done

**Review by vadmeste — DISMISSED** [review]
LGTM & tested

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by donatello — DISMISSED** [review]
I'm seeing some issue when I repeatedly run:

```
$ ./mc --insecure admin heal --recursive --json sslm1
{"status":"success","type":"summary","objects_scanned":0,"objects_healed":0,"items_scanned":0,"items_healed":0,"size":0,"duration":64}
```

I have 5 small objects in a bucket called test. But they dont seem to get scanned.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by donatello — COMMENTED** [review]
Force stopping does seem to work - however it looks like force stop always reports as success regardless of whether a heal is running or not.

**Comment by harshavardhana** [comment]
> Force stopping does seem to work - however it looks like force stop always reports as success regardless of whether a heal is running or not.

Yes there is no reason to return an error @donatello - its more like DELETE - basically i don't see anything actionable to user if we return an error.

**Review by donatello — APPROVED** [review]
Tested, LGTM

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**